graphbolt/src/cuda/common.h

```diff
@@ -88,6 +91,54 @@ inline bool is_zero<dim3>(dim3 size) {
     }                                                                 \
   }
 
+/**
+ * @brief This class is designed to handle the copy operation of a single
+ * scalar_t item from a given CUDA device pointer. Later, if the object is cast
+ * into scalar_t, the value can be read.
+ *
+ * auto num_edges = cuda::ReadScalar(indptr.data_ptr<scalar_t>() +
+ * indptr.size(0) - 1);
+ * // Perform many operations here, they will run as normal.
+ * // We finally need to read num_edges.
+ * auto indices = torch::empty(static_cast<scalar_t>(num_edges));
+ */
+template <typename scalar_t>
+struct ReadScalar {
+  ReadScalar(const scalar_t* device_ptr) : is_ready(false) {
+    pinned_scalar = torch::empty(
+        sizeof(scalar_t),
+        c10::TensorOptions().dtype(torch::kBool).pinned_memory(true));
+    auto stream = GetCurrentStream();
+    CUDA_CALL(cudaMemcpyAsync(
+        reinterpret_cast<scalar_t*>(pinned_scalar.data_ptr()), device_ptr,
+        sizeof(scalar_t), cudaMemcpyDeviceToHost, stream));
+    copy_event.record(stream);
+  }
+
+  operator scalar_t() {
+    if (!is_ready) {
+      copy_event.synchronize();
+      is_ready = true;
+    }
+    return reinterpret_cast<scalar_t*>(pinned_scalar.data_ptr())[0];
+  }
+
+ private:
+  torch::Tensor pinned_scalar;
+  at::cuda::CUDAEvent copy_event;
+  bool is_ready;
+};
+
+// This includes all integer, float and boolean types.
+#define GRAPHBOLT_DISPATCH_CASE_ALL_TYPES(...)            \
+  AT_DISPATCH_CASE_ALL_TYPES(__VA_ARGS__)                 \
+  AT_DISPATCH_CASE(at::ScalarType::Half, __VA_ARGS__)     \
+  AT_DISPATCH_CASE(at::ScalarType::BFloat16, __VA_ARGS__) \
+  AT_DISPATCH_CASE(at::ScalarType::Bool, __VA_ARGS__)
+
+#define GRAPHBOLT_DISPATCH_ALL_TYPES(TYPE, NAME, ...) \
+  AT_DISPATCH_SWITCH(TYPE, NAME, GRAPHBOLT_DISPATCH_CASE_ALL_TYPES(__VA_ARGS__))
```

Comment:
Unrelated to this PR?

Reply:
I would suggest we can create a separate PR for this code, which can be approved blindly, which is faster than adding confusion and then realize this PR did 3 things.